File: Providers/Crashlytics/AnalyticsKitCrashlyticsProvider.m
//
//  AnalyticsKitCrashlyticsProvider.m
//  AnalyticsKit
//
//  Created by Christopher Pickslay on 10/23/13.
//  Copyright (c) 2013 Two Bit Labs. All rights reserved.
//

#import <Crashlytics/Crashlytics.h>
#import "AnalyticsKitCrashlyticsProvider.h"
#import "AnalyticsKitEvent.h"
#import "AnalyticsKitTimedEventHelper.h"

@implementation AnalyticsKitCrashlyticsProvider

-(void)logScreen:(NSString *)screenName {
    CLSLog(@"screen: %@", screenName);
}

-(void)logEvent:(NSString *)event {
    CLSLog(@"event: %@", event);
    [self logEvent:event withProperties:nil];
}

-(void)logEvent:(NSString *)event withProperty:(NSString *)key andValue:(NSString *)value {
    CLSLog(@"event: %@, key: %@, value: %@", event, key, value);
    [self logEvent:event withProperties:@{key: value}];
}

-(void)logEvent:(NSString *)event withProperties:(NSDictionary *)dict {
    CLSLog(@"event: %@, properties: %@", event, dict);
    [[Crashlytics sharedInstance] logEvent:event attributes:dict];
}

-(void)logEvent:(NSString *)event timed:(BOOL)timed {
    if (timed) {
        CLSLog(@"timed event started: %@", event);
        [AnalyticsKitTimedEventHelper startTimedEventWithName:event forProvider:self];
    } else {
        [self logEvent:event];
    }
}

-(void)logEvent:(NSString *)event withProperties:(NSDictionary *)dict timed:(BOOL)timed {
    if (timed) {
        CLSLog(@"timed event started: %@ properties: %@", event, dict);
        [AnalyticsKitTimedEventHelper startTimedEventWithName:event properties:dict forProvider:self];
    } else {
        [self logEvent:event withProperties:dict];
    }
}

-(void)endTimedEvent:(NSString *)name withProperties:(NSDictionary *)dict {
    AnalyticsKitEvent *event = [AnalyticsKitTimedEventHelper endTimedEventNamed:name forProvider:self];
    if (event != nil) {
        [self logEvent:event.name withProperties:event.properties];
    }
}

-(void)logError:(NSString *)name message:(NSString *)message exception:(NSException *)exception {
    CLSLog(@"error: %@ message: %@ exception: %@", name, message, exception);
}

-(void)logError:(NSString *)name message:(NSString *)message error:(NSError *)error {
    CLSLog(@"error: %@ message: %@ exception: %@", name, message, error);
}

-(void)uncaughtException:(NSException *)exception{
    CLSLog(@"uncaught exception: %@", exception);
}

-(void)applicationDidEnterBackground {
    CLSLog(@"applicationDidEnterBackground");
}

-(void)applicationWillEnterForeground {
    CLSLog(@"applicationWillEnterForeground");
}

-(void)applicationWillTerminate {
    CLSLog(@"applicationWillTerminate");
}

@end

File: Providers/Crashlytics/AnalyticsKitCrashlyticsProvider.swift
import Foundation
import Crashlytics

public class AnalyticsKitCrashlyticsProvider: NSObject, AnalyticsKitProvider {
    fileprivate func clsLog(_ message: String) {
        CLSLogv(message, getVaList([]))
    }

    public func applicationWillEnterForeground() {}
    public func applicationDidEnterBackground() {}
    public func applicationWillTerminate() {}

    // MARK: - Log Screens

    public func logScreen(_ screenName: String) {
        clsLog("screen: \(screenName)")
        Answers.logCustomEvent(withName: "Screen - \(screenName)", customAttributes: nil)
    }

    public func logScreen(_ screenName: String, withProperties properties: [String: Any]) {
        clsLog("screen: \(screenName) properties: \(properties)")
        Answers.logCustomEvent(withName: "Screen - \(screenName)", customAttributes: properties)
    }

    // MARK: - Log Events

    public func logEvent(_ event: String) {
        clsLog("event: \(event)")
        Answers.logCustomEvent(withName: event, customAttributes: nil)
    }

    public func logEvent(_ event: String, withProperties properties: [String: Any]) {
        clsLog("event: \(event) properties: \(properties)")
        Answers.logCustomEvent(withName: event, customAttributes: properties)
    }

    public func logEvent(_ event: String, withProperty key: String, andValue value: String) {
        logEvent(event, withProperties: [key: value])
    }

    public func logEvent(_ event: String, timed: Bool) {
        logEvent(event, withProperties: [:], timed: timed)
    }

    public func logEvent(_ event: String, withProperties dict: [String: Any], timed: Bool) {
        if timed {
            clsLog("timed event started: \(event) properties: \(dict)")
            AnalyticsKitTimedEventHelper.startTimedEventWithName(event, properties: dict, forProvider: self)
            logEvent(event, withProperties: dict)
        } else {
            logEvent(event, withProperties: dict)
        }
    }

    public func endTimedEvent(_ event: String, withProperties dict: [String: Any]) {
        if let timedEvent = AnalyticsKitTimedEventHelper.endTimedEventNamed(event, forProvider: self) {
            clsLog("timed event ended: \(event) properties: \(dict)")
            logEvent(timedEvent.name, withProperties: timedEvent.properties)
        }
    }

    // MARK: - Log Errors

    public func logError(_ name: String, message: String?, properties: [String: Any]?, exception: NSException?) {
        clsLog("error: \(name) message: \(message ?? "nil") properties: \(propertiesString(for: properties) ?? "nil") exception name: \(exception?.name.rawValue ?? "nil") reason: \(exception?.reason ?? "nil")")
    }

    public func logError(_ name: String, message: String?, properties: [String: Any]?, error: Error?) {
        clsLog("error: \(name) message: \(message ?? "nil") properties: \(propertiesString(for: properties) ?? "nil") error: \(error?.localizedDescription ?? "nil")")
    }

    public func uncaughtException(_ exception: NSException) {
        clsLog("uncaught exception: \(exception.name.rawValue) reason: \(exception.reason ?? "nil")")
    }

    private func propertiesString(for properties: [String: Any]?) -> String? {
        guard let properties = properties else { return nil }
        var stringArray = [String]()
        for (key, value) in properties {
            stringArray.append("\(key): \(value)")
        }
        return "[" + stringArray.joined(separator: ", ") + "]"
    }
}
